Quick Fix Import Button

JaroslavTulach opened the pull request:
### Pull Request Description

Based on the ideas presented in the #12019 discussion this PR introduces a _quick fix button_ capable of automatically adding the missing `import` statement when a fully qualified name is used without a library being imported.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
- [x] Manually tested
- [x] Unit tests [skipped for now](https://github.com/enso-org/enso/pull/12051#issuecomment-2601925191)

JaroslavTulach committed:
Adding quick Fix Import button

JaroslavTulach committed:
Display only one icon to fit into existing layout

JaroslavTulach commented:
Let's start with following source code:
```ruby
from Standard.Base import all

main =
    file1 = Standard.Table.Row.Row
    file2 = Row
    node1 = Standard.Microsoft.Main
```
by default all three nodes yield an error:

![Missing imports](https://github.com/user-attachments/assets/28e3194b-86b0-435e-add1-f69d29a6458c)

The `Row` node shows the regular error message as present these days, because the message isn't recognized as something that could be fixed.

However there is another `Standard.Table.Row.Row` node and it offers a _quick fix_ button (instead of the copy message). Let's click it:

![One import added](https://github.com/user-attachments/assets/5239262a-0ed4-4bb9-bad9-a5cc0ea8e24d)

The source code gets a new `import Standard.Table` and the  `Standard.Table.Row.Row` node immediately resolves to a node without error. The same can be repeated with the `Standard.Microsoft.Main` node. After clicking its _Fix Import_ button, the node gets fixed. The final source code looks like this:

```ruby
from Standard.Base import all
import Standard.Table
import Standard.Microsoft

main =
    file1 = Standard.Table.Row.Row
    file2 = Row
    node1 = Standard.Microsoft.Main
```

vitvakatu commented on app/gui/src/project-view/components/GraphEditor/GraphNodeMessage.vue:
```diff
@@ -1,16 +1,44 @@
+function fixImport() {
+  const libName = containsLibraryName()
+  if (typeof libName == `string`) {
+    window.navigator.clipboard.writeText(libName)
```
I suppose it is not needed?

JaroslavTulach replied:
Removed in 80b9575e50

farmaazon commented on app/gui/src/project-view/components/GraphEditor/GraphNodeMessage.vue:
```diff
@@ -84,4 +126,7 @@ export const colorForMessageType: Record<MessageType, string> = {
 .copyButton:active {
   color: var(--color-text-inversed);
 }
+.quickFixButton:active {
+  color: var(--color-text-inversed);
+}
```
I think these two could be merged with`& > .SvgButton:active` above; I don't know why the copyButton was separate.

JaroslavTulach replied:
OK, reusing the style of copy button in dd6e07cb0c

> I don't know why the copyButton was separate.

I don't feel I should be touching the `copyButton`

farmaazon replied:
> OK, reusing the style of copy button in [dd6e07c](https://github.com/enso-org/enso/commit/dd6e07cb0cc7e55b4caf72966acb151dfcbc2e59)
> 
> > I don't know why the copyButton was separate.
> 
> I don't feel I should be touching the `copyButton`

Well, I rather meant this change:

```
diff --git a/app/gui/src/project-view/components/GraphEditor/GraphNodeMessage.vue b/app/gui/src/project-view/components/GraphEditor/GraphNodeMessage.vue
index 1acfd7ad9f..4a24a26629 100644
--- a/app/gui/src/project-view/components/GraphEditor/GraphNodeMessage.vue
+++ b/app/gui/src/project-view/components/GraphEditor/GraphNodeMessage.vue
@@ -66,7 +66,7 @@ export const colorForMessageType: Record<MessageType, string> = {
       <SvgButton
         v-if="containsLibraryName()"
         name="edit"
-        class="copyButton"
+        class="fixImportButton"
         title="Fix Import"
         @click.stop="fixImport"
       />
@@ -119,10 +119,7 @@ export const colorForMessageType: Record<MessageType, string> = {
 
   & > .SvgButton:active {
     background-color: color-mix(in oklab, black, transparent 70%);
+    color: var(--color-text-inversed);
   }
 }
-
-.copyButton:active {
-  color: var(--color-text-inversed);
-}
 </style>
```

If reusing copyButton class, it should be renamed (as it wouldn't be a class for just a copy button). But as we already have selector for "all active buttons inside toolbar", I prefer using that one.

JaroslavTulach replied:
Integrated as 6e06f3c145, thanks for the patch

Akirathan commented on app/gui/src/project-view/components/GraphEditor/GraphNodeMessage.vue:
```diff
@@ -1,16 +1,44 @@
 }>()
 
+function containsLibraryName(): string | null {
+  const prefix = 'Compile error: Fully qualified name references a library '
```
Isn't this message somewhere as constant? Seems a bit dangerous to hardcode it like this.

JaroslavTulach replied:
It may be a constant, but in Java code. We don't have a way to share constants between Java and JavaScript. However you are right. This is slightly fragile, in the long run we could try:

JaroslavTulach replied:
Let's _enhance the language server protocol_ we have. For example let's extend [Diagnostic](https://github.com/enso-org/enso/blob/develop/docs/language-server/protocol-language-server.md#diagnostic) with possible hints, like:
```js
hint: {
  type: "import",
  add: "Standard.Table"
}
```
that way we would formalize this API - would be better in the long run then depending on the text of the message.

JaroslavTulach committed:
Merge branch 'develop' into wip/jtulach/FixImport12019

JaroslavTulach committed:
Removing unnecessary lines

JaroslavTulach committed:
Let quick fix button reuse copyButton style

JaroslavTulach committed:
npx prettier

JaroslavTulach commented:
The fix import button continues to _"fix import"_ even when latest develop branch was merged in.
- I think have addressed review comments
- I haven't written any unit test - I am not sure what to copy...

Can I merge? Before today's demo?

JaroslavTulach committed:
Note in changelog

farmaazon commented:
For unit testing, the code needs to be a bit refactored... But I think they're not needed, as this is easy party and will be overhauled at some point (to not use hard-coded message, for example).

But before merging, see [my comment](https://github.com/enso-org/enso/pull/12051#discussion_r1922104310)

farmaazon committed:
Using .SvgButton:active style

JaroslavTulach commented:
Green enough. Let's get this in and claim a cooperation success.
